File: frontend/src/assets/swap.ts
import { toast, Bounce } from 'react-toastify'
import { Program, BN, AnchorProvider } from '@coral-xyz/anchor';
import { AnchorWallet } from '@solana/wallet-adapter-react';
import * as web3 from '@solana/web3.js';
// @ts-ignore
import { getAssociatedTokenAddress, TOKEN_PROGRAM_ID } from '@solana/spl-token';
import constants from './constants.ts';
import { IDL, Presale } from "./idl.ts";
import { ASSOCIATED_PROGRAM_ID } from '@coral-xyz/anchor/dist/cjs/utils/token';
import signAndSendTransaction from './signAndSendTransaction.ts';


const handleSwapTransaction = async (e: React.FormEvent<HTMLFormElement>, connection: web3.Connection, anchorWallet: AnchorWallet, isBuying: boolean, tokenAmount: string, solAmount: string) => {
    e.preventDefault();
    
    const provider = new AnchorProvider(connection, anchorWallet, {commitment: "confirmed",});
    const program = new Program<Presale>(IDL, constants.PROGRAM_ID, provider);
    
    const bATA = await getAssociatedTokenAddress(constants.TOKEN_MINT, anchorWallet.publicKey);
    
    const [presaleAuth, bump] = await web3.PublicKey.findProgramAddress(
      [constants.TOKEN_MINT.toBuffer(), constants.PROGRAM_ID.toBuffer()],
      constants.PROGRAM_ID
    );

    const pATA = await getAssociatedTokenAddress(constants.TOKEN_MINT, presaleAuth, true);

    try {
      let instruction;
      const transaction = new web3.Transaction();
      if (isBuying) {
        let buyer_balance = '0'; 

        try {
          const balance = await connection.getTokenAccountBalance(bATA);
          buyer_balance = balance.value.amount;
        } catch (error) {
          console.error('Error fetching token account balance:', error);
        }   
        if ((Number(buyer_balance) + Number(tokenAmount) * constants.TOKEN_DECIMALS) > Number(constants.MAX_TOKENS_PER_WALLET * constants.TOKEN_DECIMALS)) {
          return toast.error(`Limit of ${constants.MAX_TOKENS_PER_WALLET/constants.EXCHANGE_RATE}SOL per buyer.`, {
              position: "bottom-center",
              autoClose: 5000,
              hideProgressBar: false,
              closeOnClick: false,
              pauseOnHover: false,
              draggable: false,
              progress: undefined,
              theme: "colored",
              transition: Bounce,
            }
          );
        }
        instruction = await program.methods.buyToken(new BN(Number(solAmount) * web3.LAMPORTS_PER_SOL), bump).accounts({
          swap: {
          tokenMint: constants.TOKEN_MINT,
          buyerAssociatedTokenAccount: bATA,
          presaleAssociatedTokenAccount: pATA,
          buyer: anchorWallet.publicKey,
          vault: presaleAuth,
          presaleAuth: presaleAuth,
          rent: web3.SYSVAR_RENT_PUBKEY,
          systemProgram: web3.SystemProgram.programId,
          tokenProgram: TOKEN_PROGRAM_ID,
          associatedTokenProgram: ASSOCIATED_PROGRAM_ID,
          }
        }).instruction();
      } else {
        instruction = await program.methods.sellToken(new BN((Number(tokenAmount)) * constants.TOKEN_DECIMALS), bump).accounts({
          swap: {
          tokenMint: constants.TOKEN_MINT,
          buyerAssociatedTokenAccount: bATA,
          presaleAssociatedTokenAccount: pATA,
          buyer: anchorWallet.publicKey,
          vault: presaleAuth,
          presaleAuth: presaleAuth,
          rent: web3.SYSVAR_RENT_PUBKEY,
          systemProgram: web3.SystemProgram.programId,
          tokenProgram: TOKEN_PROGRAM_ID,
          associatedTokenProgram: ASSOCIATED_PROGRAM_ID,
          }
        }).instruction();
      }

      transaction.add(instruction);

      let transactionSignature;
      if ('phantom' in window) {
        const anyWindow: any = window;
        const phantom = anyWindow.phantom?.solana;
    
        if (phantom?.isPhantom && phantom?.publicKey?.toBuffer()[0] === anchorWallet.publicKey.toBuffer()[0]) {
          transaction.feePayer = phantom?.publicKey;

          const anyTransaction: any = transaction;
          anyTransaction.recentBlockhash = (await connection.getLatestBlockhash()).blockhash;

          transactionSignature = await signAndSendTransaction(phantom, transaction);
        } 
        else {
          transactionSignature = await provider.sendAndConfirm(transaction);
        }
      }
      else {
        transactionSignature = await provider.sendAndConfirm(transaction);
      } 
  

      console.log(`Transaction signature: ${transactionSignature}`);
      toast.success('Transaction sent.', {
        position: "bottom-center",
        autoClose: 5000,
        hideProgressBar: false,
        closeOnClick: false,
        pauseOnHover: false,
        draggable: false,
        progress: undefined,
        theme: "colored",
        transition: Bounce,
      });
    } catch (error) {
      console.error(error);
      toast.error('Transaction failed.', {
                position: "bottom-center",
                autoClose: 5000,
                hideProgressBar: false,
                closeOnClick: false,
                pauseOnHover: false,
                draggable: false,
                progress: undefined,
                theme: "colored",
                transition: Bounce,
      });
      if (error instanceof Error) {
        if (error.message.includes("SoldOut")) {
          toast.info('Not enough tokens available.', {
            position: "bottom-center",
            autoClose: 5000,
            hideProgressBar: false,
            closeOnClick: false,
            pauseOnHover: false,
            draggable: false,
            progress: undefined,
            theme: "colored",
            transition: Bounce,
          });
        } 
      }
    }
  };

  export default handleSwapTransaction;

File: smart contract/programs/contract/src/instructions/swap.rs
use {
    anchor_lang::prelude::*,
    anchor_lang::system_program,
    anchor_spl::token,
    anchor_spl::associated_token,
};

use crate::constants;

pub fn buy_token(
    ctx: Context<BuyToken>, 
    sol_amount: u64,
    bump: u8,
) -> Result<()> {
    const CORRECTION: u64 = constants::WSOL_DECIMALS / constants::TOKEN_DECIMALS;
    let token_amount = sol_amount * constants::TOKENS_PER_SOL / CORRECTION;

    if ctx.accounts.swap.presale_associated_token_account.amount < token_amount {
        return err!(Error::SoldOut)
    }

    if constants::LP_SIZE_TOKEN * constants::TOKEN_DECIMALS > (ctx.accounts.swap.presale_associated_token_account.amount - token_amount) {
        return err!(Error::SoldOut)
    }

    system_program::transfer(
        CpiContext::new(
            ctx.accounts.swap.system_program.to_account_info(),
            system_program::Transfer {
                from: ctx.accounts.swap.buyer.to_account_info(),
                to: ctx.accounts.swap.vault.to_account_info(),
            }
        ),
        sol_amount
    )?;
    msg!("SOL transferred successfully.");

    msg!("Transferring presale tokens from the vault {}...", &ctx.accounts.swap.presale_auth.key());  
    msg!("Mint: {}", &ctx.accounts.swap.token_mint.to_account_info().key());  
    msg!("From Token Address: {}", &ctx.accounts.swap.presale_associated_token_account.key());     
    msg!("To Token Address: {}", &ctx.accounts.swap.buyer_associated_token_account.key());     
    token::transfer(
        CpiContext::new_with_signer(
            ctx.accounts.swap.token_program.to_account_info(),
            token::Transfer {
                from: ctx.accounts.swap.presale_associated_token_account.to_account_info(),
                to: ctx.accounts.swap.buyer_associated_token_account.to_account_info(),
                authority: ctx.accounts.swap.presale_auth.to_account_info(),
            },
            &[&[ctx.accounts.swap.token_mint.key().as_ref(), constants::PROGRAM_ID.as_ref(), &[bump]][..]],
        ),
        token_amount
    )?;

    Ok(())
}

#[error_code]
pub enum Error {
    #[msg("Allocation of tokens for sale reached.")]
    SoldOut
}

pub fn sell_token(
    ctx: Context<SellToken>, 
    token_amount: u64,
    bump: u8,
) -> Result<()> {
    const CORRECTION: u64 = constants::WSOL_DECIMALS / constants::TOKEN_DECIMALS;
    let sol_amount = token_amount * CORRECTION / constants::TOKENS_PER_SOL;

    msg!("Transferring presale tokens to the vault {}...", &ctx.accounts.swap.presale_auth.key());  
    msg!("Mint: {}", &ctx.accounts.swap.token_mint.to_account_info().key());  
    msg!("From Token Address: {}", &ctx.accounts.swap.buyer_associated_token_account.key());     
    msg!("To Token Address: {}", &ctx.accounts.swap.presale_associated_token_account.key());     
    token::transfer(
        CpiContext::new(
            ctx.accounts.swap.token_program.to_account_info(),
            token::Transfer {
                from: ctx.accounts.swap.buyer_associated_token_account.to_account_info(),
                to: ctx.accounts.swap.presale_associated_token_account.to_account_info(),
                authority: ctx.accounts.swap.buyer.to_account_info(),
            }
        ),
        token_amount
    )?;

    system_program::transfer(
        CpiContext::new_with_signer(
            ctx.accounts.swap.system_program.to_account_info(),
            system_program::Transfer {
                from: ctx.accounts.swap.vault.to_account_info(),
                to: ctx.accounts.swap.buyer.to_account_info(),
            },
            &[&[ctx.accounts.swap.token_mint.key().as_ref(), constants::PROGRAM_ID.as_ref(), &[bump]][..]],
        ),
        sol_amount
    )?;
    msg!("SOL transferred successfully.");

    Ok(())
}


#[derive(Accounts)]
pub struct Swap<'info> { 

#[account(
    mut
)]
pub token_mint: Box<Account<'info, token::Mint>>,

#[account(
    init_if_needed,
    payer = buyer,
    associated_token::mint = token_mint,
    associated_token::authority = buyer,
)]
pub buyer_associated_token_account: Box<Account<'info, token::TokenAccount>>,

#[account(
    mut,
    associated_token::mint = token_mint,
    associated_token::authority = vault,
)]
pub presale_associated_token_account: Box<Account<'info, token::TokenAccount>>,

#[account(mut)]
pub buyer: Signer<'info>,

/// CHECK:
#[account(
    mut,
    seeds = [token_mint.key().as_ref(), constants::PROGRAM_ID.as_ref()],
    bump
)]
pub vault: UncheckedAccount<'info>,
pub presale_auth: SystemAccount<'info>,

pub rent: Sysvar<'info, Rent>,
pub system_program: Program<'info, System>,
pub token_program: Program<'info, token::Token>,
pub associated_token_program: Program<'info, associated_token::AssociatedToken>,
}

#[derive(Accounts)]
pub struct BuyToken<'info> {
    pub swap: Swap<'info>,

}

#[derive(Accounts)]
pub struct SellToken<'info> {
    pub swap: Swap<'info>,

}